File: vortexdbg-api/src/main/java/com/vortexdbg/arm/context/BaseRegisterContext.java
package com.vortexdbg.arm.context;

import com.vortexdbg.Emulator;
import com.vortexdbg.arm.backend.Backend;
import com.vortexdbg.pointer.VortexdbgPointer;

public abstract class BaseRegisterContext extends AbstractRegisterContext implements RegisterContext {

    protected final Emulator<?> emulator;
    private final int firstArgReg;
    private final int regArgCount;

    BaseRegisterContext(Emulator<?> emulator, int firstArgReg, int regArgCount) {
        this.emulator = emulator;
        this.firstArgReg = firstArgReg;
        this.regArgCount = regArgCount;
    }

    @Override
    public VortexdbgPointer getPointerArg(int index) {
        if (index < regArgCount) {
            int reg = firstArgReg + index;
            return VortexdbgPointer.register(emulator, reg);
        }

        VortexdbgPointer sp = getStackPointer();
        return sp.getPointer((long) (index - regArgCount) * emulator.getPointerSize());
    }

    @Override
    public int getIntByReg(int regId) {
        Backend backend = emulator.getBackend();
        Number number = backend.reg_read(regId);
        return number.intValue();
    }

    @Override
    public long getLongByReg(int regId) {
        Backend backend = emulator.getBackend();
        Number number = backend.reg_read(regId);
        return number.longValue();
    }

}

File: vortexdbg-api/src/main/kotlin/com/vortexdbg/arm/context/BaseRegisterContext.kt
package com.vortexdbg.arm.context

import com.vortexdbg.Emulator
import com.vortexdbg.arm.backend.Backend
import com.vortexdbg.pointer.VortexdbgPointer

abstract class BaseRegisterContext internal constructor(
    protected val emulator: Emulator<*>,
    private val firstArgReg: Int,
    private val regArgCount: Int
) : AbstractRegisterContext(), RegisterContext {

    override fun getPointerArg(index: Int): VortexdbgPointer? {
        if (index < regArgCount) {
            val reg = firstArgReg + index
            val p = VortexdbgPointer.register(emulator, reg)
            return if (p.peer == 0L) null else p
        }

        val sp = getStackPointer()
        val p = sp.getPointer((index - regArgCount).toLong() * emulator.getPointerSize())
        return if (p == null || p.peer == 0L) null else p
    }

    override fun getIntByReg(regId: Int): Int {
        val backend: Backend = emulator.getBackend()
        val number: Number = backend.reg_read(regId)
        return number.toInt()
    }

    override fun getLongByReg(regId: Int): Long {
        val backend: Backend = emulator.getBackend()
        val number: Number = backend.reg_read(regId)
        return number.toLong()
    }

}
